libraries/Microsoft.Bot.Builder.Dialogs/Prompts/Prompt.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Prompts;
+using static Microsoft.Bot.Builder.Prompts.PromptValidatorEx;
+
+namespace Microsoft.Bot.Builder.Dialogs
+{
+    /// <summary>
+    /// Basic configuration options supported by all prompts.
+    /// </summary>
+    public abstract class Prompt<T> : Control where T : PromptResult
+    {
+        public Prompt()
+        {
+        }
+
+        public override bool HasDialogContinue => true;
+
+        public override bool HasDialogResume => false;
+
+        protected abstract Task OnPrompt(DialogContext dc, PromptOptions options, bool isRetry);
+
+        protected abstract Task<T> OnRecognize(DialogContext dc, PromptOptions options);
+
+        public override async Task DialogBegin(DialogContext dc, object dialogArgs)
```

Comment:
Why this? Strong type and avoid the cast. 

public override new async Task DialogBeging(DialogContext dc, PromptOptions dialogArgs)

Reply:
yeah

---
In reply to: [181198215](https://github.com/Microsoft/botbuilder-dotnet/pull/408#discussion_r181198215) [](ancestors = 181198215)

Reply:
The abstract Dialog class defines the method signature and it doesn't know anything about Prompts or PromptOptions. Prompts are derived from Dialog.